numpy/lib/shape_base.py

```diff
@@ -73,66 +76,66 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
-    indlist = list(range(nd))
-    indlist.remove(axis)
-    i[axis] = slice(None, None)
-    outshape = asarray(arr.shape).take(indlist)
-    i.put(indlist, ind)
-    res = func1d(arr[tuple(i.tolist())], *args, **kwargs)
-    #  if res is a number, then we have a smaller output array
-    if isscalar(res):
-        outarr = zeros(outshape, asarray(res).dtype)
-        outarr[tuple(ind)] = res
-        Ntot = product(outshape)
-        k = 1
-        while k < Ntot:
-            # increment the index
-            ind[-1] += 1
-            n = -1
-            while (ind[n] >= outshape[n]) and (n > (1-nd)):
-                ind[n-1] += 1
-                ind[n] = 0
-                n -= 1
-            i.put(indlist, ind)
-            res = func1d(arr[tuple(i.tolist())], *args, **kwargs)
-            outarr[tuple(ind)] = res
-            k += 1
-        return outarr
-    else:
-        res = asanyarray(res)
-        Ntot = product(outshape)
-        holdshape = outshape
-        outshape = list(arr.shape)
-        outshape[axis] = res.size
-        outarr = zeros(outshape, res.dtype)
-        outarr = res.__array_wrap__(outarr)
-        outarr[tuple(i.tolist())] = res
-        k = 1
-        while k < Ntot:
-            # increment the index
-            ind[-1] += 1
-            n = -1
-            while (ind[n] >= holdshape[n]) and (n > (1-nd)):
-                ind[n-1] += 1
-                ind[n] = 0
-                n -= 1
-            i.put(indlist, ind)
-            res = asanyarray(func1d(arr[tuple(i.tolist())], *args, **kwargs))
-            outarr[tuple(i.tolist())] = res
-            k += 1
-        if res.shape == ():
-            outarr = outarr.squeeze(axis)
-        return outarr
+
+    # arr, with the iteration axis at the end
+    dims_in = list(range(nd))
+    inarr_view = transpose(arr, dims_in[:axis] + dims_in[axis+1:] + [axis])
+
+    # indices
+    inds = ndindex(inarr_view.shape[:nd-1])
+
+    # invoke the function on the first item
+    ind0 = next(inds)
+    res = func1d(inarr_view[ind0], *args, **kwargs)
+    res = asanyarray(res)
+
+    # insert as many axes as necessary to create the output
+    outshape = arr.shape[:axis] + res.shape + arr.shape[axis+1:]
+    outarr = zeros(outshape, res.dtype)
+    # matrices call reshape in __array_prepare__, and this is apparently ok!
+    if not isinstance(res, matrix):
+        outarr = res.__array_prepare__(outarr)
+    if outshape != outarr.shape:
+        raise ValueError('__array_prepare__ should not change the shape of the resultant array')
+
+    # outarr, with inserted dimensions at the end
+    # this means that outarr_view[i] = func1d(inarr_view[i])
+    dims_out = list(range(outarr.ndim))
+    outarr_view = transpose(outarr, dims_out[:axis] + dims_out[axis+res.ndim:] + dims_out[axis:axis+res.ndim])
+
+    # save the first result
+    outarr_view[ind0] = res
+    for ind in inds:
+        outarr_view[ind] = asanyarray(func1d(inarr_view[ind], *args, **kwargs))
```

Comment:
I agree that keeping the same memory layout should be preferred in general. But in this case, keeping the same memory layout requires jumping through some hoops, and the resulting code is much less obvious. So I would lean towards changing it.